### PR TITLE
Componentization Pt. 2

### PR DESCRIPTION
**Pull request by Guillar1**

### CODE REVIEW

File: packages/itwin/grouping-mapping-widget/src/widget/components/CustomCalculationTable.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/CustomCalculationTable.tsx:138]: Copy-paste error.

File: packages/itwin/grouping-mapping-widget/src/widget/components/CustomCalculationTable.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/CustomCalculationTable.tsx:131]: These toolbars should be a separate component. That should reduce code duplication between different property types.

File: packages/itwin/grouping-mapping-widget/src/widget/components/viewerUtils.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/viewerUtils.tsx:98]: Doesn't KeySet already contain ids? This looks like a duplication of data. If there is no required accessor in KeySet, maybe it is worth to contact the `@itwin/presentation-common` lib maintainers?

**Review by MindaugasButkus — COMMENTED** [review]
There is quite a lot of code duplication between different property components. I'll continue reviewing once the code duplication comments are resolved.

File: packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyTable.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyTable.tsx]: Done.

File: packages/itwin/grouping-mapping-widget/src/widget/components/CustomCalculationTable.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/CustomCalculationTable.tsx]: Fixed

File: packages/itwin/grouping-mapping-widget/src/widget/components/CustomCalculationTable.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/CustomCalculationTable.tsx]: Separated it to its own component.

File: packages/itwin/grouping-mapping-widget/src/widget/components/viewerUtils.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/viewerUtils.tsx]: It does, but it is not the complete set of ID's needed to render the element. That is where ```getHiliteSet``` comes into play. I cache the result of ```getHiliteSet``` even though it states that 
```
Note: The provider caches result of the last request, so subsequent requests
for the same input doesn't cost.
```
because it still makes a network request and there is a considerable slow down.

File: packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyAction.tsx:80]: Yes, I do agree that this should be done in a separate PR. These leaks have been a issue for a while now.

File: packages/itwin/grouping-mapping-widget/src/widget/hooks/useFetchData.ts
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/hooks/useFetchData.ts]: Done.

File: packages/itwin/grouping-mapping-widget/src/widget/hooks/useFetchData.ts
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/hooks/useFetchData.ts:31]: Saving this for another PR.

File: packages/itwin/grouping-mapping-widget/src/widget/components/CustomCalculationAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/CustomCalculationAction.tsx:109]: Querying of properties should be done in parallel. There is no point doing them in sequence.

File: packages/itwin/grouping-mapping-widget/src/widget/components/CustomCalculationTable.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/CustomCalculationTable.tsx:53]: The state name is too verbose. `showDeleteModal` would be enough. There is only one delete modal in this component.

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyMenu.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/PropertyMenu.tsx:127]: If error occurs in any of the above async operations the widget will be stuck at loading state.

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyMenu.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/PropertyMenu.tsx:116]: Will the user be able to modify properties while the elements within a group are being loaded? Not all properties require visualizing all group elements.

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyMenu.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/PropertyMenu.tsx:141]:
```suggestion
            onChange={() => setShowGroupColor(!showGroupColor)}
```

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyMenu.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/PropertyMenu.tsx:191]: Group's InformationPanel should be a separate component.

File: packages/itwin/grouping-mapping-widget/src/widget/components/groupsHelpers.ts
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/groupsHelpers.ts:25]: Get a distinct list of group queries and get all ids in parallel.

File: packages/itwin/grouping-mapping-widget/src/widget/components/groupsHelpers.ts
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/groupsHelpers.ts:38]: Get all ids in parallel for all distinct group queries.

**Review by MindaugasButkus — CHANGES_REQUESTED** [review]

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTableToolbar.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTableToolbar.tsx:27]: Inconsistent use of quotes. You have all 3 supported quote types, when 1 would be enough. You use backticks for a string above, but you do not do any string interpolation there.

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyNameCell.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/PropertyNameCell.tsx:8]: Why no `export`?

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTableToolbar.tsx
Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTableToolbar.tsx]: Any reasoning why you kept destructuring of props?

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyMenu.tsx
Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/components/PropertyMenu.tsx:144]: Nothing changed, but marking this as resolved to unblock this PR.

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTableToolbar.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTableToolbar.tsx:27]: Linter should have picked this up. Fixed it.

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyNameCell.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/PropertyNameCell.tsx:8]: Didn't push this yet. Fixed

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyNameCell.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/PropertyNameCell.tsx]: Switched to interface.

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyMenu.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/PropertyMenu.tsx]: Yes.

File: packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyTable.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyTable.tsx:40]: Why did you change from `useMemo` to `useCallback` here? Also, this is not `columns` now but rather `columnsFactory`.

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTable.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTable.tsx:67]: Looks like columns will be recreated for every re-render. Earlier columns were stored in a `useMemo`.

File: packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyTable.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyTable.tsx:111]: The signature of this function would look nicer if it only accepted `propertyId`. I see no point requiring `iModelId` and `accessToken` if `groupId` and `mappingId` are captured.

Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyTable.tsx:111]: I'm not against it. I will just have to bring `const { getAccessToken, iModelId } = useGroupingMappingApiConfig();` to every table to capture it. Is that ok?

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTable.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTable.tsx:67]: `useMemo` is used to memoize a value (like a function's result), while `useCallback` is used to memoize a function itself, so that it is not recreated on each render of the component. So this should not happen if the columns callback is wrapped in `useCallback`.

File: packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyTable.tsx
Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyTable.tsx:111]: Yes. The signature would make more sense.

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTable.tsx
Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTable.tsx:67]: But you are assigning a value to `columns` prop, not a callback. Every re-render this `columns` function will be executed and the return value passed into the `columns` props of `Table` component.

File: packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyTable.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyTable.tsx]: Updated signature.

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTable.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTable.tsx]: Oh true, nice catch. My mistake. I have fixed this now.

File: packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyTable.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/CalculatedPropertyTable.tsx]: I agree. `columnsFactory` is a lot clearer.

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTable.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTable.tsx:51]: What do you memoize here? Factory, or the data that is returned by factory? Name does not match the contents.

**Review by MindaugasButkus — APPROVED** [review]

File: packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTable.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTable.tsx:51]: Made it clearer. `memoizedColumns`

Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTable.tsx:51]: You probably didn't push your changes yet.

Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/PropertyTable.tsx]: Huh, it errored out. Second try worked.

**Comment by Guillar1** [comment]
Not going to merge until after QA.

File: packages/itwin/grouping-mapping-widget/src/widget/components/Grouping.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/Grouping.tsx:253]: It's pretty hard to understand where different parts of this ternary operator start/end. Maybe moving the whole `DropdownMenu` into a separate component would help.

File: packages/itwin/grouping-mapping-widget/src/widget/components/Grouping.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/Grouping.tsx:240]: moved it to a separate file.

File: packages/itwin/grouping-mapping-widget/src/widget/components/Grouping.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/Grouping.tsx:253]: Extracted the ```DropDownmenu```/```Button``` to its own file.

**Pull request merged**
